### ISSUE TITLE
Correct way to consume multiple files

### ISSUE THREAD

**Issue by JoaoAparicio** [google/grain#708]
What is the correct way to consume multiple file?

I've implemented a `RandomAccessDataSource` which opens and pre-processes each file.

To me the benefit of grain is that I don't have to manually implement batching or sequences (for RNN), and grain knows how to shape the arrays accordingly. If the solution to consuming multiple files is that I have one `RandomAccessDataSource` per file, wrap those in `DataLoader`s, and then for loop over the `DataLoader`s/files, then I lose the batching benefit - I'm effectively having to implement batching.

Is there a way of telling a `DataLoader` "your data is these N `RandomAccessDataSource`"?

**Comment by sirfz** [comment]
Funny, I just tested out grain for an experiment I'm working on and needed something like this (sample from multiple data sources). I found [`ConcatenateMapDataset`](https://google-grain.readthedocs.io/en/latest/autoapi/mix/index.html#mix.ConcatenateMapDataset) which I believe should work for you. It's not exposed via `grain.python` so I had to import it like this:

```python
from grain._src.python.dataset.transformations.mix import ConcatenateMapDataset

data_sources: list[grain.MapDataset]  # list[grain.RandomAccessDataSource] works too btw
dataset = ConcatenateMapDataset(data_sources)
loader = grain.DataLoader(data_source=dataset, ...)
```

I used a custom `grain.Sampler` which randomly samples from each dataset with weighted probabilties.

**Comment by rajasekharporeddy** [comment]
Hi @JoaoAparicio,

Thanks for raising this question! You don't need to loop over separate `DataLoader`s per file—Grain supports combining multiple `RandomAccessDataSource` instances into a single logical dataset so that shuffling and batching happen seamlessly across file boundaries.

You can wrap each file's `RandomAccessDataSource` with `grain.MapDataset.source()`, concatenate them with `grain.MapDataset.concatenate()`, and apply batching/shuffling directly:

```python
import grain

class MyFileDataSource(grain.sources.RandomAccessDataSource):
    def __init__(self, data):
        self._data = data
    def __len__(self):
        return len(self._data)
    def __getitem__(self, idx):
        return self._data[idx]

# Suppose each file has its own RandomAccessDataSource
file_sources = [
    MyFileDataSource([1, 2, 3]),
    MyFileDataSource([4, 5]),
    MyFileDataSource([6, 7, 8, 9]),
]

# 1. Wrap each source and concatenate into a single MapDataset
combined_ds = grain.MapDataset.concatenate(
    [grain.MapDataset.source(s) for s in file_sources]
)

# 2. Apply transformations (batches will seamlessly cross file boundaries)
batched_ds = combined_ds.shuffle(seed=42).batch(batch_size=4)

for batch in batched_ds:
    print(batch)
# [9 7 8 4]
# [1 6 2 3]
# [5]
```

Closing this issue as resolved—please feel free to re-open if you have any further questions!
